### PR TITLE
FEAT: updated jupyterhub module 

### PR DESCRIPTION
**Pull request by umeshkumhar**
Few updates:
- remove kubectl provider dependency from jupyter-on-gke
- Moved IAP yamls into helm chart
- Removed multiple hardcoded values into variables.
- Added auto-fetch for backend-service after first apply, (just re-apply terraform, and it will update oauth audience with proper ID)
- added terraform.tfvars.example for user input reference, use that as well to perform testing of jupyter-on-gke

Testing scenarios:
- First duplicate terraform.tfvars.example to terraform.tfvars. and update its values.
- run terraform apply, if add_auth=true, this will configure everything related to IAP.. but jupyterhub will be installed with wrong audience URI (because its not created yet).
- Once all done, again rerun terraform apply, and this time, it will identify backend_service_id and will update helm chart with new audience, that will upgrade jupyter-hub deployment.
- Fetch domain/ip from ingress or managedcertificate.

### CODE REVIEW

File: applications/ray/main.tf
Comment (imreddy13) [inline applications/ray/main.tf:1]: Should there be a new folder under applications for jupyter?

Shouldn't jupyter-on-gke live under /modules and be installed by the jupyter application?

File: jupyter-on-gke/jupyterhub.tf [hunk 1]
```diff
@@ -26,111 +36,100 @@ provider "helm" {
...
     })
   ]
+  depends_on = [module.iap_auth]
+}
 
-  depends_on = [
-    kubernetes_namespace.namespace
-  ]
+
+# Need to re-apply: fetch service_id from deployed service
```
Comment (imreddy13) [inline jupyter-on-gke/jupyterhub.tf:112]: Is this still true with the helm chart, does it need a re-apply?

Can we also have an updated README with instructions to install?

File: jupyter-on-gke/variables.tf [hunk 2]
```diff
@@ -15,13 +15,12 @@
 variable "namespace" {
   type        = string
   description = "Kubernetes namespace where resources are deployed"
-  default     = "<your user name>"
+  default     = "umesh"
```
Comment (imreddy13) [inline jupyter-on-gke/variables.tf:18]: revert this?

File: jupyter-on-gke/variables.tf [hunk 3]
```diff
@@ -75,18 +62,18 @@ variable "url_domain_name" {
 variable "support_email" {
   type        = string
   description = "Email for users to contact with questions about their consent"
-  default     = "<Support email>"
+  default     = "umeshkumhar@google.com"
```
Comment (imreddy13) [inline jupyter-on-gke/variables.tf:65]: same

File: jupyter-on-gke/allowlist [hunk 4]
```diff
@@ -1,3 +1,2 @@
-user:example@google.com
-user:example2@google.com
-group:example-groupd@google.com
\ No newline at end of file
+allUsers
+user:umeshkumhar@google.com
```
Comment (imreddy13) [inline jupyter-on-gke/allowlist:2]: revert this?

File: jupyter-on-gke/variables.tf (hunk 2 above)
Comment (umeshkumhar) [reply jupyter-on-gke/variables.tf:18]: this wont be used as user inputs.. its just a placeholder.

will anyway update in new PR

File: jupyter-on-gke/variables.tf (hunk 3 above)
Comment (umeshkumhar) [reply jupyter-on-gke/variables.tf:65]: same here

File: jupyter-on-gke/allowlist (hunk 4 above)
Comment (umeshkumhar) [reply jupyter-on-gke/allowlist:2]: same here

File: infrastructure/tfvars_examples/private-standard-gke-with-new-network.platform.tfvars [hunk 5]
```diff
@@ -142,7 +142,7 @@ all_node_pools_oauth_scopes = [
 ]
 
 all_node_pools_labels = {
-  "cloud.google.com/gke-profile" = "ray"
+  "gke-profile" = "ray"
```
Comment (imreddy13) [inline infrastructure/tfvars_examples/private-standard-gke-with-new-network.platform.tfvars:145]: should this be "ai-on-gke"?

File: jupyter-on-gke/allowlist (hunk 4 above)
Comment (imreddy13) [reply jupyter-on-gke/allowlist:2]: sorry, lost the original comment. should we remove your ldap?

File: jupyter-on-gke/jupyterhub.tf [hunk 6]
```diff
@@ -26,111 +36,100 @@ provider "helm" {
...
-  name            = "jupyterhub"
-  repository      = "https://jupyterhub.github.io/helm-chart"
-  chart           = "jupyterhub"
-  namespace       = var.namespace
-  cleanup_on_fail = "true"
+  name             = "jupyterhub"
+  repository       = "https://jupyterhub.github.io/helm-chart"
+  chart            = "jupyterhub"
+  namespace        = var.namespace
+  create_namespace = !contains(data.kubernetes_all_namespaces.allns.namespaces, var.namespace)
```
Comment (imreddy13) [inline jupyter-on-gke/jupyterhub.tf:96]: Where is this var being used?

File: jupyter-on-gke/terraform.tfvars.example [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+####    APPLICATIONS
+#######################################################
+
+## JupyterHub variables
+namespace = "jupyter-ns"
+
+# Jupyterhub with IAP
+add_auth = true
+brand = "projects/494975018659/brands/494975018659"
+support_email = "umeshkumhar@google.com"
```
Comment (imreddy13) [inline jupyter-on-gke/terraform.tfvars.example:31]: fix this?

File: jupyter-on-gke/terraform.tfvars.example [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+#######################################################
+####    APPLICATIONS
+#######################################################
+
+## JupyterHub variables
+namespace = "jupyter-ns"
+
+# Jupyterhub with IAP
+add_auth = true
+brand = "projects/494975018659/brands/494975018659"
```
Comment (imreddy13) [inline jupyter-on-gke/terraform.tfvars.example:30]: same

File: jupyter-on-gke/jupyterhub.tf (hunk 6 above)
Comment (umeshkumhar) [reply jupyter-on-gke/jupyterhub.tf:96]: used by helm, if namespace creation is required or not.

File: infrastructure/tfvars_examples/private-standard-gke-with-new-network.platform.tfvars (hunk 5 above)
Comment (umeshkumhar) [reply infrastructure/tfvars_examples/private-standard-gke-with-new-network.platform.tfvars:145]: updated.

File: jupyter-on-gke/terraform.tfvars.example (hunk 7 above)
Comment (umeshkumhar) [reply jupyter-on-gke/terraform.tfvars.example:31]: updated.

File: jupyter-on-gke/terraform.tfvars.example (hunk 8 above)
Comment (umeshkumhar) [reply jupyter-on-gke/terraform.tfvars.example:30]: updated.

**Review by imreddy13 — APPROVED** [review]

**Pull request merged**